pkg/services/dashboards/database/database.go

```diff
@@ -962,17 +965,236 @@ func (d *dashboardStore) FindDashboards(ctx context.Context, query *dashboards.F
+		derivedCtx := context.WithoutCancel(ctx)
+		go func() {
+			start := time.Now()
+			results, err := d.findDashboards(derivedCtx, query)
+			if err != nil {
+				d.log.Info("Alternative search query failed", "error", err)
+				return
+			}
+			d.log.Info("Alternative search query", "time", time.Since(start), "results", len(results))
+
+			expected := <-resc
+
+			expectedUIDs, resultUIDs, n, m := map[string]struct{}{}, map[string]struct{}{}, 0, 0
+			for _, hit := range expected {
+				expectedUIDs[hit.UID] = struct{}{}
+			}
+			for _, hit := range results {
+				resultUIDs[hit.UID] = struct{}{}
+				if _, ok := expectedUIDs[hit.UID]; !ok {
+					d.log.Info("Alternative search query mismatch", "unexpected", hit.Title, "uid", hit.UID)
+					n++
+				}
+			}
+			for _, hit := range expected {
+				if _, ok := resultUIDs[hit.UID]; !ok {
+					d.log.Info("Alternative search query mismatch", "missed", hit.Title, "uid", hit.UID)
+					m++
+				}
+			}
+			if n+m > 0 {
+				d.log.Info("Alternative search query got different results", "matched", (len(expected)+len(results)-n-m)/2, "unexpected", n, "missed", m)
+			}
+		}()
+	}
+
+	start := time.Now()
 	err = d.store.WithDbSession(ctx, func(sess *db.Session) error {
 		return sess.SQL(sql, params...).Find(&res)
 	})
-
+	if d.features.IsEnabled(ctx, featuremgmt.FlagSearchAlt) {
+		d.log.Info("Original search query", "time", time.Since(start), "results", len(res))
+	}
 	if err != nil {
 		return nil, err
 	}
 
+	resc <- res
 	return res, nil
 }
 
+// searchHit represents a single matching row from the search query below.
+type searchHit struct {
+	OrgID       int64  `xorm:"org_id"`
+	ID          int64  `xorm:"id"` // this probably should be deprecated but it will break dashboard_tag table for now
+	Kind        string `xorm:"kind"`
+	UID         string `xorm:"uid"`
+	Title       string `xorm:"title"`
+	Term        string `xorm:"term"`
+	FolderTitle string `xorm:"folder_title"`
+	FolderUID   string `xorm:"folder_uid"`
+}
+
+// searchData is an input data for the search query template below.
+type searchData struct {
+	SQLBuilder
+	Text    string
+	OrgID   int64
+	UserID  int64
+	TeamIDs []int64
+	Role    string
+	Limit   int64
+
+	Dialect        migrator.Dialect
+	HasSplitScopes bool
+}
+
+// SQLBuilder is a more-or-less reusable SQL query builder that is based on
+// text/template. SQL is not a composable language really, so having all
+// subqueries/partials/filters in one place helps to better understand the whole
+// query.
+// SQLBuilder is expected to be embedded into the data struct passed into the
+// template. In a template one can use {{.Arg .Foo}} and it will replace that
+// with a "?" placeholder appending the value of .Foo to the Params slice.
+type SQLBuilder struct {
+	Params []any
+}
+
+func (sb *SQLBuilder) Arg(v any) (sql string) { sb.Params = append(sb.Params, v); return "?" }
+
+var searchSQL = template.Must(template.New("").Parse(`
```

Comment:
Using templating here is such a neat idea. It's great to be able to have an overview of the whole query like this.

One of my hopes when I first heard about this approach was to have an overview in which the meaningful subunits of the query are abstracted away (in their own variables). One example here would be the various selects which take part in a union.

One advantage of doing this would be to be able to understand the important parts of the query at one glance rather than having to read through blocks of sql. Another advantage is being able to more easily document the various parts of the query.

Definitely only a suggestion, but it would be great to see the query broken down in this way.